Sources/FoundationEssentials/Subprocess/Subprocess+AsyncBytes.swift

```diff
@@ -0,0 +1,110 @@
+        @inline(__always) static var bufferSize: Int {
+            16384
+        }
+        public typealias Element = UInt8
+
+        @_nonSendable
+        public struct Iterator: AsyncIteratorProtocol {
+            public typealias Element = UInt8
+
+            private let fileDescriptor: FileDescriptor
+            private var buffer: [UInt8]
+            private var currentPosition: Int
+            private var finished: Bool
+
+            internal init(fileDescriptor: FileDescriptor) {
+                self.fileDescriptor = fileDescriptor
+                self.buffer = []
+                self.currentPosition = 0
+                self.finished = false
+            }
+
+            private mutating func reloadBufferAndNext() async throws -> UInt8? {
+                if self.finished {
+                    return nil
+                }
+                try Task.checkCancellation()
+                do {
+                    self.buffer = try await self.read(
+                        from: self.fileDescriptor,
+                        maxLength: AsyncBytes.bufferSize)
+                    self.currentPosition = 0
+                    if self.buffer.count < AsyncBytes.bufferSize {
+                        self.finished = true
+                    }
+                } catch {
+                    self.finished = true
+                    throw error
+                }
+                return try await self.next()
+            }
+
+            public mutating func next() async throws -> UInt8? {
+                if currentPosition < buffer.count {
+                    let value = buffer[currentPosition]
+                    self.currentPosition += 1
+                    return value
+                }
+                return try await self.reloadBufferAndNext()
+            }
+
+            private func read(from fileDescriptor: FileDescriptor, maxLength: Int) async throws -> [UInt8] {
+                return try await withCheckedThrowingContinuation { continuation in
+                    DispatchIO.read(
+                        fromFileDescriptor: fileDescriptor.rawValue,
+                        maxLength: maxLength,
+                        runningHandlerOn: .main
+                    ) { data, error in
+                        guard error == 0 else {
+                            continuation.resume(
+                                throwing: POSIXError(
+                                    .init(rawValue: error) ?? .ENODEV)
+                            )
+                            return
+                        }
+                        continuation.resume(returning: Array(data))
+                    }
+                }
+            }
+        }
+
+        private let fileDescriptor: FileDescriptor
+
+        init(fileDescriptor: FileDescriptor) {
+            self.fileDescriptor = fileDescriptor
+        }
+
+        public func makeAsyncIterator() -> Iterator {
+            return Iterator(fileDescriptor: self.fileDescriptor)
```

Comment:
We must enforce here that only a single iterator is created ever and `fatalError` on subsequent iterator creation otherwise there might be concurrent reads on the FD.